### PR TITLE
Update for various things - 0.6.0

### LINKED ISSUE #124: Attribute 'queue' (used to get a copy of the data) missing from Queue in persistqueue
In the standard Queue from python we can access the attribute queue to get a copy of the data, like so:

```
>>> from queue import Queue
>>> q = Queue()
>>> q.put('a')
>>> q.put('b')
>>> print(q)
<queue.Queue object at 0x01CEA310>
>>> print(q.queue)
deque(['a', 'b'])
>>> for elem in q.queue:
...   print(elem)
a
b
>>> q.get()
'a'
>>> q.get()
'b'
```

Notice that q.queue is an iterable object that contains each element of the queue, and since it's a copy of the data, accessing it doesn't modify the actual queue (when you use get(), you still get the first element). In contrast, the actual queue 'q' is not iterable and if you try to print it, it will not show you the data inside it.

This .queue attribute is very useful for logging purposes and also in automated tests to do assertions with the data inside the queue.

I want to replace my current queue with a persistent queue, but I ran into the following problem when trying to access that .queue attribute

```
>>> q.queue
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Queue' object has no attribute 'queue'
```

**Do you have any suggestions on how I could access the "persistent queue" data without actually modifying the queue for logging and automated testing purposes?**

Thanks!

### LINKED ISSUE #126: Clean DB
As I queue jobs and complete them, the database doesn't decrease in size that I can tell in my tests.  I just appears to grow and grow, saving the history.  Once a job is complete, I want it gone as my jobs are large in size and I don't want it to store any anything.  How do I keep the DB clean so it only contains pending jobs, not historical ones?

### LINKED ISSUE #129: nack option - end of queue
It would be nice if there was an option for nack that sent it to the end of the queue.  Right now I'm doing an ack and doing another put, which may be what you would do internally, but anyway, figured I'd ask.  I'm running into the issue of a job failing and then it just continues to retry the same job over and over with a nack, failing each time.  After so many tries, I'll probably do a ack_failed, but for now I don't want it to hold up other jobs.

### LINKED ISSUE #144: qsize() reports negative value
Using the SQLiteAckQueue multithreaded and I'm trying to get a count on the queue size using qsize(), but sometimes I'll see negative numbers. I think that the issue is the `_unack_cache`  is not included in the total queue size, which I think it probably should be as it hasn't yet received an `ack/task_done`  So in a mutlithreaded situation where you have 10 get calls, you're going to report a queue size 10 less than what actually exists, since those jobs haven't been completed.

This also causes an ever changing qsize as things can be `get`/`nack` regularly, so the size keeps going up and down as the state changes.  I think it should only change the size when a new item is added `put` or an item is removed  `ack/ack_failed/task_done`.  I don't think a `get` or `nack` should change the reported size.

### LINKED ISSUE #146: __str__ for Queue feature request
I'm probably dumb, but for debugging I'd like to always see the entire queue.  Can you add an __str__ method to Queue please

### LINKED ISSUE #149: Queue Item Metadata
If I do a `queue.get()`, it gives me the data in the queue, but there is metadata with items such as the id, timestamp, and status.  I'd like to be able to read the timestamp.  Is there an easy way to do that when I get() a queue item?  Thanks

### LINKED ISSUE #150: Why is  _MAX_ACKED_LENGTH Hardcoded?
In this code, _MAX_ACKED_LENGTH is hardcoded:
https://github.com/peter-wangxu/persist-queue/blob/5ce10afcfcd81afd46a9eae32ba085d8ae900240/persistqueue/sqlackqueue.py#L35

Why is it hardcoded to 1000? I need to remove about 10 to 100 thousand acked entries each time I run this, is there any reason I shouldn't change _MAX_ACKED_LENGTH to 100_000 and then run that?

### PR DESCRIPTION
**Pull request by imidoriya**
I don't normally like to make large PR requests, but since I started addressing one issue, I figured I'd update the rest for the things I'm using.  Unit tests included and all changes preserve backward compatibility. What does this fix / change?  

- Resolves https://github.com/peter-wangxu/persist-queue/pull/131, Resolves  https://github.com/peter-wangxu/persist-queue/pull/82, and Resolves #150, Resolves #126 by adding several options to `clear_acked_data()`
    - `max_delete` (defaults to 1000): This is the LIMIT.  How many items to delete.
    - `keep_latest` (defaults to 1000): This is the OFFSET.  How many recent items to keep
    - `clear_ack_failed` (defaults to False): Clears the ack_failed as the original only clears ack.

- Resolves #149 and Resolves #129 by returning the `id` on put along with an ack, nack, ack_failed on SQLiteAckQueue.  It also adds options to `get()`
    - `raw` (defaults to False): Returns the metadata along with the record, which includes the id (`pqid`) and timestamp.  On the SQLiteAckQueue, the raw results can be ack, nack, ack_failed similar to the normal return.  
    -  `id` (defaults to None): Accepts an `id` as returned above or a raw item containing `pqid`.  Will select the item based on the row id.
    -  `next_in_order` SQLiteAckQueue (defaults to False): Requires the `id` attribute.  This option tells the SQLiteAckQueue get the next item based on ID, not the first available.  This allows the user to get, nack, get, nack and progress down the queue, instead of continuing to get the same nack'd item over again.  Prior to this, you had to move the item to the back of the queue to get the next item.

`raw` example:
```py
q.put('val1')
d = q.get(raw=True)
print(d)
{'pqid': 1, 'data': 'val1', 'timestamp': 1616719225.012912}
q.ack(d)
```

`next_in_order` example:
 ```py
q.put("val1")
q.put("val2")
q.put("val3")
item = q.get()
id = q.nack(item)
item = q.get(id=id, next_in_order=True)
print(item)
val2
```

- Resolves #144 by adding a new size query `active_size()` for SQLiteAckQueue that includes the unacked_cache length.  I would have replaced `qsize()` but didn't want to mess with backward compatibility.
- Resolves #124 and Resolves #146 by added a new queue query `queue()` that returns the queue as an List[Dict].  SQLiteAckQueue includes status.
```py
print(q.queue())
[{'id': 1, 'data': 'var1, 'timestamp': 1616718505.389051}, {'id': 2, 'data': 'var2', 'timestamp': 1616718543.814977}, {'id': 3, 'data': 'var3', 'timestamp': 1616718543.816642}]
```
Added SQLiteAckQueue `update()`. I added this as I sometimes want to update an item when I nack it, such as adding a retry count.
```py
qid = q.put("val1")
q.update(id=qid, item="val2")
item = q.get(id=qid)
print(item)
val2
```

### CODE REVIEW

**Comment by peter-wangxu** [comment]
@imidoriya thanks for the commit, really awesome!

there are various fixes/features which are deadly desired by the community.
Before I merge the great PR, please make the ut tests and pep8 pass.

Thanks
Peter

**Comment by imidoriya** [comment]
@peter-wangxu will do on the pep8.  On the ui tests, the ones that fail are failing for me with your master branch as well.  They're `serializer` issues with `queue.py` that I don't really know how to fix except to just remove the serializer as I don't understand that piece of code.  If you could take a look and offer a suggestion, happy to implement it.

Example:
```
======================================================================
ERROR: test_put_maxsize_reached (persistqueue.tests.test_queue.PersistTest)
Test that maxsize reached.
----------------------------------------------------------------------
TypeError: test_put_maxsize_reached() missing 1 required positional argument: 'serializer'
```

File: persistqueue/tests/test_sqlackqueue.py [hunk 1]
```diff
@@ -290,6 +290,55 @@ def test_put_0(self):
         d = q.get(block=False)
         self.assertIsNotNone(d)
 
+    def test_get_id(self):
+        q = SQLiteAckQueue(path=self.path)
+        q.put("val1")
+        val2 = q.put("val2")
+        q.put("val3")
+        item = q.get(item=val2)
```
Comment (peter-wangxu) [inline persistqueue/tests/test_sqlackqueue.py:298]: use`(id=val2)`

File: persistqueue/tests/test_sqlackqueue.py [hunk 2]
```diff
@@ -290,6 +290,55 @@ def test_put_0(self):
...
+        self.assertEqual(val2, 2)
+        # item should get val2
+        self.assertEqual(item, 'val2')
+
+    def test_get_next_in_order(self):
+        q = SQLiteAckQueue(path=self.path)
+        val1 = q.put("val1")
+        q.put("val2")
+        q.put("val3")
+        item = q.get(item=val1, next_in_order=True)
```
Comment (peter-wangxu) [inline persistqueue/tests/test_sqlackqueue.py:309]: Ditto

File: persistqueue/tests/test_sqlqueue.py [hunk 3]
```diff
@@ -234,14 +234,33 @@ def test_json_serializer(self):
...
         q.put(0)
         d = q.get(block=False)
         self.assertIsNotNone(d)
 
+    def test_get_id(self):
+        q = SQLiteQueue(path=self.path)
+        q.put("val1")
+        val2 = q.put("val2")
+        q.put("val3")
+        item = q.get(item=val2)
```
Comment (peter-wangxu) [inline persistqueue/tests/test_sqlqueue.py:250]: use`(id=val2)`

**Review by peter-wangxu — CHANGES_REQUESTED** [review]
Looks great except some nits.

File: persistqueue/tests/test_sqlqueue.py (hunk 3 above)
Comment (imidoriya) [reply persistqueue/tests/test_sqlqueue.py:250]: Ah, dang.. I had renamed those to make it simpler. Guess that's the point of unit tests.  Doesn't work if you don't rerun them. lol

**Comment by imidoriya** [comment]
Added an update() function for SQLiteAckQueue.

**Comment by imidoriya** [comment]
I now see that to run the unittests, you use `nose2` not `python3 -m unittest`.  That allows the `queue.py` to pass the serializer.  Learning something new. :)  From what I can tell, the issue failing the check seems to be building the greenlet wheel on py34.  Not sure how to fix that or if py34 is still needed for CI.

**Comment by peter-wangxu** [comment]
let's ignore the windows py34 issue.

I am fine with the whole change, no schema change that may break backward compatibility.
Let's see if there's any concern raised by other users.


Thanks
Peter

**Comment by kstratis** [comment]
Does this require any documentation updates?

**Comment by imidoriya** [comment]
> Does this require any documentation updates?

Yes, probably need to add that somewhere.  The repo doesn't have the Wiki enabled, so I was looking at the readme, but I didn't want to complicate things up for new users.  Just not sure the best way to present the options.  @peter-wangxu what do you think?  You want me to put these options in the readme or do you want to enable the wiki to detail the more advanced features of persist-queue?

**Comment by peter-wangxu** [comment]
> > Does this require any documentation updates?
> 
> Yes, probably need to add that somewhere. The repo doesn't have the Wiki enabled, so I was looking at the readme, but I didn't want to complicate things up for new users. Just not sure the best way to present the options. @peter-wangxu what do you think? You want me to put these options in the readme or do you want to enable the wiki to detail the more advanced features of persist-queue?

I created a [wiki page](https://github.com/peter-wangxu/persist-queue/wiki/Advanced-usage-of-queues), you can add more detailed usage there.

I think adding the interfaces added in this PR is still needed at the readme.

**Comment by imidoriya** [comment]
> I created a [wiki page](https://github.com/peter-wangxu/persist-queue/wiki/Advanced-usage-of-queues), you can add more detailed usage there.
> 
> I think adding the interfaces added in this PR is still needed at the readme.

Info added to the readme.  I wasn't able to update the Wiki.  Looks like permission might not be set to public?  Check under the Settings-> Options -> Features -> Restrict editing to collaborators only (should be unchecked if you want to make it publicly editable)

**Comment by peter-wangxu** [comment]
I added you as collaborator now

**Review by peter-wangxu — APPROVED** [review]
LGTM

**Comment by kstratis** [comment]
Thank you both for the excellent work! ❤️

**Pull request merged**